### DISCUSSION TITLE
For first previewers

### DISCUSSION THREAD

**Discussion by langningchen** [langningchen/XMOJ-Client#7] [General]
I you have any questions, feel free to discuss here!

**Comment by PythonSmall-Q** [comment]
你知道对接有多累吗，同时和两个人说话

**Comment by PythonSmall-Q** [comment]
然后还要一边写代码

**Reply by boomzero** [reply]
+1

**Comment by PythonSmall-Q** [comment]
网还特别差，真**无语

**Reply by PythonSmall-Q** [reply]
倒苦水来了

**Comment by langningchen** [comment]
我也是……

**Comment by boomzero** [comment]
Qusetion, isn't python supposed to be python3?

**Reply by PythonSmall-Q** [reply]
好像本质上不一样，看你装的是哪一个

**Comment by boomzero** [comment]
Could not find PYTHON_EXECUTABLE using the following names: python

**Reply by PythonSmall-Q** [reply]
python windows自带，python3自己要装

**Comment by langningchen** [comment]
我是按照 `python` 查找的

https://github.com/langningchen/XMOJ-Client/blob/4824ef1265824388a572060adbc046b0a882e81a/Backend/CMakeLists.txt#L8

**Reply by boomzero** [reply]
> 我是按照 `python` 查找的
> 
> https://github.com/langningchen/XMOJ-Client/blob/4824ef1265824388a572060adbc046b0a882e81a/Backend/CMakeLists.txt#L8

Okay

**Comment by langningchen** [comment]
有什么方法能同时查找 python 和 python3 吗？

**Reply by boomzero** [reply]
just search for python3, I think

**Reply by boomzero** [reply]
In linux/macOS, python3 exists

**Reply by boomzero** [reply]
And on macOS, python doesn't exist

**Comment by langningchen** [comment]
但是在 Windows 中，默认只有 python ，大部分用户都是使用 Windows

**Reply by boomzero** [reply]
Environment variable "CC" is not set...

**Comment by langningchen** [comment]
`CC` 这个很奇怪，应该是一个上游的 BUG ，但是我没有提出来

**Reply by boomzero** [reply]
我一开始可以的啊, 怎么没set了

**Comment by boomzero** [comment]
Ohh am I supposed to use run.sh

**Comment by langningchen** [comment]
https://github.com/Mbed-TLS/mbedtls/blob/ea7fff481e5bd17247cc915f8c322d239d4ebe4f/scripts/ecp_comb_table.py#L30

而这个文件是在编译的时候会运行，但是 `MbedTLS` 没有对 `CC` 环境变量检查

**Reply by boomzero** [reply]
> https://github.com/Mbed-TLS/mbedtls/blob/ea7fff481e5bd17247cc915f8c322d239d4ebe4f/scripts/ecp_comb_table.py#L30
> 
> 而这个文件是在编译的时候会运行，但是 `MbedTLS` 没有对 `CC` 环境变量检查

```
-- Performing Test HAVE_STRUCT_TIMEVAL - Success
-- Performing Test HAVE_CLOCK_GETTIME_MONOTONIC_RAW
-- Performing Test HAVE_CLOCK_GETTIME_MONOTONIC_RAW - Success
-- Enabled features: SSL IPv6 UnixSockets libz AsynchDNS Largefile alt-svc HSTS NTLM HTTPS-proxy threadsafe
-- Enabled protocols: DICT FILE FTP FTPS GOPHER GOPHERS HTTP HTTPS IMAP IMAPS IPFS IPNS LDAP LDAPS MQTT POP3 POP3S RTSP SCP SFTP SMB SMBS SMTP SMTPS TELNET TFTP
-- Enabled SSL backends: mbedTLS
-- Configuring incomplete, errors occurred!
make: Makefile: No such file or directory
make: *** No rule to make target `Makefile'.  Stop.
```
What happened...

**Comment by langningchen** [comment]
即使你运行了 `Run.sh` 后面也会出很多奇怪的错，比如 MbedTLS 找不到库什么的……

**Reply by boomzero** [reply]
找不到makefile...

**Comment by langningchen** [comment]
> ```
> -- Performing Test HAVE_STRUCT_TIMEVAL - Success
> -- Performing Test HAVE_CLOCK_GETTIME_MONOTONIC_RAW
> -- Performing Test HAVE_CLOCK_GETTIME_MONOTONIC_RAW - Success
> -- Enabled features: SSL IPv6 UnixSockets libz AsynchDNS Largefile alt-svc HSTS NTLM HTTPS-proxy threadsafe
> -- Enabled protocols: DICT FILE FTP FTPS GOPHER GOPHERS HTTP HTTPS IMAP IMAPS IPFS IPNS LDAP LDAPS MQTT POP3 POP3S RTSP SCP SFTP SMB SMBS SMTP SMTPS TELNET TFTP
> -- Enabled SSL backends: mbedTLS
> -- Configuring incomplete, errors occurred!
> make: Makefile: No such file or directory
> make: *** No rule to make target `Makefile'.  Stop.
> ```
> 
> What happened...

我没见过呢……很奇怪

**Reply by PythonSmall-Q** [reply]
库没装全

**Reply by boomzero** [reply]
然后我再运行:

```
/bin/zsh /Users/chenqi/Desktop/OSS/XMOJ-Client/Build.sh
CMake Error at Backend/CMakeLists.txt:5 (message):
  Environment variable "CC" is not set


-- Configuring incomplete, errors occurred!
make: Makefile: No such file or directory
make: *** No rule to make target `Makefile'.  Stop.
```

**Reply by PythonSmall-Q** [reply]
https://github.com/XMOJ-Script-dev/XMOJ-Script/pull/603 @boomzero

**Reply by PythonSmall-Q** [reply]
CC这个env没设置

**Comment by langningchen** [comment]
我这里没有遇到过那个 Makefile 的问题……

**Reply by PythonSmall-Q** [reply]
那是你环境设置全

**Reply by boomzero** [reply]
> 我这里没有遇到过那个 Makefile 的问题……

It's a cmake error

**Reply by boomzero** [reply]
And it's about CC

**Comment by boomzero** [comment]
CMake Error at Backend/CMakeLists.txt:53 (add_subdirectory):
  The source directory

    /Users/chenqi/Desktop/OSS/XMOJ-Client/Backend/Libraries/SQLiteCpp

  does not contain a CMakeLists.txt file.

**Comment by boomzero** [comment]
I need to install it, I suppose

**Comment by langningchen** [comment]
我发现一个小问题

**Comment by langningchen** [comment]
我有一个地方写错了 `git` 写成了 `Git`

**Reply by boomzero** [reply]
哦我说怎么不build呢

**Comment by langningchen** [comment]
> I need to install it, I suppose

这个他应该会自动安装的，但是因为我写错了，所以没有安装……

**Comment by langningchen** [comment]
https://github.com/langningchen/XMOJ-Client/blob/4824ef1265824388a572060adbc046b0a882e81a/Backend/CMakeLists.txt#L28

这里，你先把前面的 Git 手动改一下

**Reply by boomzero** [reply]
https://github.com/boomzero/XMOJ-Client/actions/runs/10208721474/job/28245604742

**Comment by langningchen** [comment]
他在编译的时候很可能出错

**Reply by boomzero** [reply]
Windows failed...

**Comment by boomzero** [comment]
```
clang: warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]
/home/runner/work/XMOJ-Client/XMOJ-Client/Backend/Classes/Problems.cpp:207:34: error: use of undeclared identifier 'Database'
    DATABASE::SQL_DATA SQLData = Database.Select("Problems", {}, {{"ID", std::to_string(ID)}});
                                 ^
/home/runner/work/XMOJ-Client/XMOJ-Client/Backend/Classes/Problems.cpp:213:13: error: use of undeclared identifier 'Database'
            Database.Delete("Problems", {{"ID", std::to_string(ID)}});
            ^
```

**Comment by langningchen** [comment]
欸，不是……

**Comment by langningchen** [comment]
哦，sorry，这个好像是我之前改了一半的……

**Comment by langningchen** [comment]
Sorry, see: https://github.com/langningchen/XMOJ-Client/commit/4824ef1265824388a572060adbc046b0a882e81a

**Comment by langningchen** [comment]
我本来想统一一下数据库，因为我有的地方使用的数据库是全局变量，有的地方是传参，后来好像只改了一半……

**Comment by boomzero** [comment]
So I just revert https://github.com/langningchen/XMOJ-Client/commit/4824ef1265824388a572060adbc046b0a882e81a ?

**Comment by langningchen** [comment]
@boomzero I have one new commit, check it!

**Comment by langningchen** [comment]
The commit can be compiled on my machine

**Reply by boomzero** [reply]
https://github.com/boomzero/XMOJ-Client/commit/cc254a8f29dccef153c2c725645807eacd24904c

**Reply by boomzero** [reply]
...

**Comment by langningchen** [comment]
https://github.com/langningchen/XMOJ-Client/actions/runs/10208843073

**Comment by langningchen** [comment]
对的

**Comment by langningchen** [comment]
> `gmake[2]: *** No rule to make target 'Backend/Libraries/mbedtls/library/libmbedtls.a', needed by 'Backend/XMOJ-Client-Backend'.  Stop.` 

是一个常见错误

**Comment by langningchen** [comment]
我也不知道怎么解决……

**Comment by langningchen** [comment]
我每次都是直接绕过问题……因为这个问题我看了半天都没找到 Solution ，可能因为我太菜了

**Comment by boomzero** [comment]
clang failed, but gcc is still building

**Comment by langningchen** [comment]
GCC success

**Comment by boomzero** [comment]
Success

**Comment by langningchen** [comment]
No!

**Reply by boomzero** [reply]
?

**Comment by langningchen** [comment]
It isn't success

**Reply by boomzero** [reply]
Why?

**Comment by langningchen** [comment]
Because the default cmake does not make the target XMOJ-Client, it makes the curl... 

> `[100%] Built target curl`

**Reply by boomzero** [reply]
...

**Comment by langningchen** [comment]
We should use `Build.sh`

**Comment by langningchen** [comment]
https://github.com/langningchen/XMOJ-Client/actions/runs/10208921066

**Comment by langningchen** [comment]
@boomzero 这才是真正的成功 https://github.com/langningchen/XMOJ-Client/actions/runs/10208921066/job/28246115921

> `[ 92%] Built target XMOJ-Client-Frontend`
> `[100%] Built target XMOJ-Client-Backend`
> `[100%] Built target XMOJ-Client`

**Reply by boomzero** [reply]
Upload asset?

**Comment by boomzero** [comment]
Clang is angry with you...

**Comment by langningchen** [comment]
About the asset, you can add it in your fork, I will go to write some OI problems...

**Comment by PythonSmall-Q** [comment]
奇淫小技巧：用chatgpt审核PR

**Comment by PythonSmall-Q** [comment]
学习microsoft

**Comment by langningchen** [comment]
@boomzero https://github.com/langningchen/XMOJ-Client/actions/runs/10209042097 上传好了

**Comment by boomzero** [comment]
Okay, don't have linux-x86 though

**Comment by langningchen** [comment]
Windows 出错是这个：

> `2024-08-02T02:36:43.1718377Z C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(755,5): error MSB8084: Invalid structured output from 'CL.exe': Cannot parse JsonRpc notification: 'Cannot transcode invalid UTF-8 JSON text to UTF-16 string.'. [D:\a\XMOJ-Client\XMOJ-Client\Build\Backend\Libraries\curl\lib\libcurl_object.vcxproj]`

**Reply by boomzero** [reply]
> Windows 出错是这个：
> 
> > `2024-08-02T02:36:43.1718377Z C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(755,5): error MSB8084: Invalid structured output from 'CL.exe': Cannot parse JsonRpc notification: 'Cannot transcode invalid UTF-8 JSON text to UTF-16 string.'. [D:\a\XMOJ-Client\XMOJ-Client\Build\Backend\Libraries\curl\lib\libcurl_object.vcxproj]`

I am trying to fix macOS

**Comment by langningchen** [comment]
在 MacOS 上面能不能用 gcc g++ ? clang 好像不支持 -s 编译选项，他可能是另一个静态选项

**Reply by boomzero** [reply]
可以

**Reply by boomzero** [reply]
https://github.com/boomzero/XMOJ-Client/actions/runs/10209151499/job/28246724563

**Comment by boomzero** [comment]
```
  cmake --build /Users/runner/work/XMOJ-Client/XMOJ-Client/Build --config  -j 8 --target XMOJ-Client
  shell: /bin/bash -e {0}
CMake Error: Invalid value used with --config
Usage: cmake --build <dir>             [options] [-- [native-options]]
```

**Comment by boomzero** [comment]
?

**Comment by langningchen** [comment]
你的 Actions 配置有误

**Reply by boomzero** [reply]
fixed

**Comment by langningchen** [comment]
我刚才手贱把我的 frpc 关掉了…… 所以我现在连不上去了…… 我已经编译好的在远程，所以我现在改不了了……

**Reply by boomzero** [reply]
> 我刚才手贱把我的 frpc 关掉了…… 所以我现在连不上去了…… 我已经编译好的在远程，所以我现在改不了了……

...

**Comment by PythonSmall-Q** [comment]
对了，为什么我unsubscribe还会收到通知

**Comment by boomzero** [comment]
md不行

**Comment by boomzero** [comment]
去搞OI去了
